File: sparkcql/src/main/scala/gov/va/sparkcql/common/Stopwatch.scala
package gov.va.sparkcql.common

import scala.concurrent.duration._

class Stopwatch {
  val start: Long = System.nanoTime()
  var end: Long = 0

  def stop() = {
    end = System.nanoTime()
  }

  def lapse: Duration = {
    if (end == 0) {
      return Duration.fromNanos(System.nanoTime() - start)
    } else {
      return Duration.fromNanos(end - start)
    }
  }

  override def toString: String = {
    val duration = lapse
    val HH = duration.toHours
    val MM = duration.toMinutes
    val SS = duration.toSeconds
    val MS = duration.toMillis
    return s"${HH}h:${MM}m:${SS}s:${MS}ms"
  }
}

object Stopwatch {
  def apply(): Stopwatch = {
    return new Stopwatch()
  }
}

File: sparkcql-core/src/main/java/gov/va/sparkcql/diagnostic/Stopwatch.java
package gov.va.sparkcql.diagnostic;

import java.time.Duration;

public class Stopwatch {
  
    private Long startTime = System.nanoTime();
    private Long endTime = 0L;

    public void stop() {
        this.endTime = System.nanoTime();
    }

    public Duration getLapse() {
        if (endTime == 0L) {
            return Duration.ofNanos(System.nanoTime() - startTime);
        } else {
            return Duration.ofNanos(endTime - startTime);
        }
    }

    public String toString() {
        var duration = getLapse();
        var HH = duration.toHours();
        var MM = duration.toMinutes();
        var SS = duration.toSeconds();
        var MS = duration.toMillis();
        return String.format("%dh:%dm:%ds:%dms", HH, MM, SS, MS);
    }

    public void sleep(long milliseconds) {
        try {
            Thread.sleep(milliseconds);
        } catch (InterruptedException e) {
        }
    }

    public void print() {
        System.out.println(toString());
    }
}